### PR TITLE
API-1835: Don't use Update in removestaleconditionscontroller

### PR DESCRIPTION
**Pull request by bertinatto**
Proof: https://github.com/openshift/cluster-authentication-operator/pull/729

Manual testing done:

1. Introduced the removestaleconditions controller in https://github.com/openshift/gcp-pd-csi-driver-operator and marked the **ManagementStateDegraded** condition to be removed.
2. Removed the managementstate controller from the operator.
3. Watched the removestaleconditions controller remove the condition correctly.
4. Added the managementstate controller back to the operator and watched both controller fighting.

### CODE REVIEW

File: pkg/operator/staleconditions/remove_stale_conditions.go [hunk 1]
```diff
@@ -2,50 +2,70 @@ package staleconditions
...
 	"github.com/openshift/library-go/pkg/controller/factory"
 	"github.com/openshift/library-go/pkg/operator/events"
 	"github.com/openshift/library-go/pkg/operator/v1helpers"
 )
 
 type RemoveStaleConditionsController struct {
-	conditions     []string
-	operatorClient v1helpers.OperatorClient
+	controllerInstanceName string
+	conditions             []string
```
Comment (p0lyn0mial) [inline pkg/operator/staleconditions/remove_stale_conditions.go:16]: please rename to `conditionTypes` or even `conditionTypesToRemove`

File: pkg/operator/staleconditions/remove_stale_conditions.go [hunk 2]
```diff
@@ -2,50 +2,70 @@ package staleconditions
...
 		WithSync(c.sync).
 		WithInformers(operatorClient.Informer()).
 		ToController(
-			"RemoveStaleConditionsController", // don't change what is passed here unless you also remove the old FooDegraded condition
+			c.controllerInstanceName,
 			eventRecorder.WithComponentSuffix("remove-stale-conditions"),
 		)
 }
 
 func (c RemoveStaleConditionsController) sync(ctx context.Context, syncContext factory.SyncContext) error {
```
Comment (p0lyn0mial) [inline pkg/operator/staleconditions/remove_stale_conditions.go:41]: would you mind adding some unit tests for the sync function ?

File: pkg/operator/staleconditions/remove_stale_conditions.go [hunk 3]
```diff
@@ -2,50 +2,70 @@ package staleconditions
...
-			v1helpers.RemoveOperatorCondition(&status.Conditions, condition)
+	_, operatorStatus, _, err := c.operatorClient.GetOperatorState()
+	if err != nil {
+		return err
+	}
+
+	var removedCount int
+	jsonPatch := jsonpatch.New()
+	for i, existingCondition := range operatorStatus.Conditions {
+		for _, conditionToRemove := range c.conditions {
```
Comment (p0lyn0mial) [inline pkg/operator/staleconditions/remove_stale_conditions.go:50]: please rename to `conditionTypeToRemove`

File: pkg/operator/staleconditions/remove_stale_conditions.go [hunk 4]
```diff
@@ -2,50 +2,70 @@ package staleconditions
...
+			if existingCondition.Type != conditionToRemove {
+				continue
+			}
+			removeAtIndex := i
+			if !jsonPatch.IsEmpty() {
+				removeAtIndex = removeAtIndex - removedCount
+			}
+			jsonPatch.WithRemove(
+				fmt.Sprintf("/status/conditions/%d", removeAtIndex),
+				jsonpatch.NewTestCondition(fmt.Sprintf("/status/conditions/%d/type", removeAtIndex), conditionToRemove),
```
Comment (p0lyn0mial) [inline pkg/operator/staleconditions/remove_stale_conditions.go:60]: is there an e2e test that we could rely on ?

**Review by p0lyn0mial — COMMENTED** [review]
lgtm

File: pkg/operator/staleconditions/remove_stale_conditions.go (hunk 4 above)
Comment (bertinatto) [reply pkg/operator/staleconditions/remove_stale_conditions.go]: I'm not aware of any

File: pkg/operator/staleconditions/remove_stale_conditions.go (hunk 2 above)
Comment (bertinatto) [reply pkg/operator/staleconditions/remove_stale_conditions.go:41]: Added some unit tests (bases on the node status controller)

**Pull request merged**